File: server/game-server/internal-module/src/main/java/com/example/model/GameInstance.java
package com.example.model;

import java.util.*;

public class GameInstance {
    private GameState gameState;
    private Integer ID;
    private HashMap<Integer, PlayerData> players = new HashMap();

    public GameInstance(GameState gameState, Integer ID, HashMap<Integer, PlayerData> players) {
        this.gameState = gameState;
        this.ID = ID;
        if (players == null) {
            players = new HashMap();
        }

        this.players = players;
    }

    public boolean addPlayer(PlayerData player) {
        if (player != null && !this.players.containsKey(player.getPublicID())) {
            return this.players.put(player.getPublicID(), player) == null;
        } else {
            return false;
        }
    }

    public boolean updatePlayer(PlayerData player) {
        if (player != null && this.players.containsKey(player.getPublicID())) {
            return this.players.replace(player.getPublicID(), player) != null;
        } else {
            return false;
        }
    }

    public boolean removePlayer(Integer playerPublicID) {
        if (!this.players.containsKey(playerPublicID)) {
            return false;
        } else {
            return this.players.remove(playerPublicID) != null;
        }
    }

    public boolean isEmpty() {
        return this.players.isEmpty();
    }

    public boolean join() {
        System.out.println("\nGameInstance.join() called ");
        this.setGameState(GameState.GAME_LOBBY);
        return true;
    }

    public boolean start() {
        System.out.println("\nGameInstance.start() called ");
        this.setGameState(GameState.GAME_STARTING);
        return true;
    }

    public boolean enter() {
        System.out.println("\nGameInstance.entering() called ");
        this.setGameState(GameState.ENTERING_GAME);
        return true;
    }

    public boolean run() {
        System.out.println("\nGameInstance.run() called ");
        this.setGameState(GameState.IN_GAME);
        return true;
    }

    public boolean close() {
        System.out.println("\nGameInstance.close() called ");
        this.setGameState(GameState.CLOSING);
        return true;
    }

    public Integer getID() {
        return this.ID;
    }

    protected void setID(Integer iD) {
        this.ID = iD;
    }

    public GameState getGameState() {
        return this.gameState;
    }

    public void setGameState(GameState newState) {
        System.out.println("GameInstance.setGameState() newState = " + newState);
        this.gameState = newState;
    }

    public void setPlayers(HashMap<Integer, PlayerData> players) {
        this.players = players;
    }

    public Collection<PlayerData> getAllPlayers() {
        return this.players.values();
    }

    public PlayerData getPlayer(Integer playerID) {
        return (PlayerData)this.players.get(playerID);
    }

    public PlayerData getPlayer(String playerName) {
        Iterator var3 = this.players.values().iterator();

        while(var3.hasNext()) {
            PlayerData player = (PlayerData)var3.next();
            if (player.getName().equals(playerName)) {
                return player;
            }
        }

        return null;
    }

    public String toString() {
        return "\n{ '_commnent' : 'GAME','ID' : " + this.ID + ", " + "'gameState' : " + this.gameState + ", " + this.getAllPlayers() + " } ";
    }

    public static void main(String[] args) {
        System.out.println("Running GameInstance.main()");
        System.out.println("End GameInstance.main()");
    }

    public static enum GameState {
        GAME_LOBBY,
        GAME_STARTING,
        ENTERING_GAME,
        IN_GAME,
        CLOSING,
        UNKNOWN;

        private GameState() {
        }
    }
}

File: UnrealClient/Source/UnrealClient/model/GameInstance.cpp

#include "GameInstance.h"
#include "PlayerData.h"

using namespace std;


GameInstance::GameInstance()
{
}

GameInstance::GameInstance(GameState gameState, int ID, 
			unordered_map<int, PlayerData> &players)
{
	this->gameState = gameState;
	this->gameID = ID.value();

	// create new players map or set players map
	// 
	//if (players.empty())
	if (players == null)
	{
		throw new Exception("players map cannot be null");
		//this->players = unordered_map<int, PlayerData>();
	}
	else
	{
		this->players = players;
		this->gameOwner = (vector<>(players.values()))[0];
	}
}


// NOTE: this version calls other constructor to set most attributes
GameInstance::GameInstance(GameState gameState, int ID, 
			unordered_map<int, PlayerData> &players, PlayerData gameOwner)
					: GameInstance(gameState, ID, players)
{

	//if (gameOwner == nullptr)
	if (gameOwner == NULL)
	{
		throw runtime_error("gameOwner passed in is NULL");
	}

	// NOTE: this should proably be just the playerID
	this->gameOwner = gameOwner;

	// NOTE: commenting out. don't think we want to pass gameName this way
	/*
	// set game name
	 if ((static_pointer_cast<TemplePlayerData>(gameOwner))->gameName != "")
	{
		this->gameName = (gameOwner)->gameName;
	}
	else
	{
		this->gameName = "TempleRun-" + ID.value();
	}*/


}

string GameInstance::getGameName()
{
	return gameName;
}

GameInstance GameInstance::setGameName(const string gameName)
{
	this->gameName = gameName;
	return shared_from_this();
}

PlayerData GameInstance::getGameOwner()
{
	return gameOwner;
}

void GameInstance::setGameOwner(PlayerData gameOwner)
{
	this->gameOwner = gameOwner;
}

bool GameInstance::addPlayer(PlayerData player)
{

	//if (player == nullptr)
	if (player == NULL)
	{
		throw runtime_error("Null Input Exception");
	}

	return this->players.emplace(player->getPublicID(), player) == nullptr;
}

bool GameInstance::updatePlayer(PlayerData player)
{
	if (player == nullptr)
	{
		throw runtime_error("Null Input Exception");
	}
	if (players.find(player->getPublicID()) == players.end())
	{
		return false;
	}

	return this->players.replace(player->getPublicID(), player) != nullptr;
}

bool GameInstance::removePlayer(int playerPublicID)
{
	if (this->players.find(playerPublicID) == this->players.end())
	{
		return false;
	}

	return this->players.erase(playerPublicID) != nullptr;
}


//ORIGINAL LINE: @JsonIgnore public boolean isEmpty()
bool GameInstance::isEmpty()
{
	return this->players.empty();
}

//ORIGINAL LINE: @JsonIgnore public boolean isFull()
bool GameInstance::isFull()
{
	return this->players.size() >= GameDesignVars::MAX_PLAYERS_PER_GAME;
}

bool GameInstance::join()
{
	this->setGameState(GameState::GAME_LOBBY);
	return true;
}

bool GameInstance::start()
{
	this->setGameState(GameState::GAME_STARTING);
	return true;
}

bool GameInstance::enter()
{
	this->setGameState(GameState::ENTERING_GAME);
	return true;
}

bool GameInstance::run()
{
	this->setGameState(GameState::IN_GAME);
	return true;
}

bool GameInstance::close()
{
	this->setGameState(GameState::CLOSING);
	return true;
}

int GameInstance::getGameID()
{
	return this->gameID;
}

void GameInstance::setGameID(int iD)
{
	this->gameID = iD;
}

GameInstance::GameState GameInstance::getGameState()
{
	return this->gameState;
}

void GameInstance::setGameState(GameState newState)
{
	this->gameState = newState;
}

PlayerData GameInstance::getPlayer(int playerID)
{
	return this->players[playerID];
}

unordered_map<int, PlayerData> & GameInstance::getPlayers()
{
	return players;
}

string GameInstance::toString()
{
	return  "gameID=" + gameID + 
			", gameName=" + gameName + 
			", gameOwner=" + gameOwner->getName() + 
			", gameState=" + gameState;

	/*
	return "\n{ "
		+ "'_commnent' : 'GAME',"
		+ "'ID' : " + gameID + ", "
		+ "'gameState' : " + gameState + ", "
		+ getAllPlayers()
		+ " } ";
	*/

}

bool GameInstance::equals(any o)
{
	if (this == o)
	{
		return true;
	}
	if (!(o.type() == typeid(GameInstance)))
	{
		return false;
	}

	GameInstance game = any_cast<GameInstance>(o);
	if (gameID != game->gameID)
	{
		return false;
	}
	if (gameState != game->gameState)
	{
		return false;
	}

	return true;
}


public int GameInstance::hashCode() {
	int result = gameState != null ? gameState.hashCode() : 0;
	result = 31 * result + gameID;
	result = 31 * result + (players != null ? players.hashCode() : 0);
	return result;
}



void GameInstance::setCustomGameName(const string gameName)
{
	this.gameName = gameName;
}


